bfvmm/include/intrinsics/x86/intel/apic/lapic.h

```diff
@@ -55,41 +55,113 @@ namespace ia32_apic_base
+        { return set_bit(msr, from); }
+
+        inline void disable()
+        { _write_msr(addr, clear_bit(_read_msr(addr), from)); }
+
+        inline auto disable(value_type msr)
+        { return clear_bit(msr, from); }
 
         inline void dump(int level, std::string *msg = nullptr)
-        { bfdebug_subnhex(level, name, get(), msg); }
+        { bfdebug_subbool(level, name, is_enabled(), msg); }
     }
 
-    namespace state
+    namespace enable_x2apic
     {
-        constexpr const auto mask = 0xC00ULL;
+        constexpr const auto mask = 0x0000000000000400ULL;
         constexpr const auto from = 10ULL;
-        constexpr const auto name = "state";
+        constexpr const auto name = "enable_x2apic";
+
+        inline auto is_enabled()
+        { return is_bit_set(_read_msr(addr), from); }
+
+        inline auto is_enabled(value_type msr)
+        { return is_bit_set(msr, from); }
+
+        inline auto is_disabled()
+        { return is_bit_cleared(_read_msr(addr), from); }
+
+        inline auto is_disabled(value_type msr)
+        { return is_bit_cleared(msr, from); }
+
+        inline void enable()
+        { _write_msr(addr, set_bit(_read_msr(addr), from)); }
+
+        inline auto enable(value_type msr)
+        { return set_bit(msr, from); }
+
+        inline void disable()
+        { _write_msr(addr, clear_bit(_read_msr(addr), from)); }
+
+        inline auto disable(value_type msr)
+        { return clear_bit(msr, from); }
+
+        inline void dump(int level, std::string *msg = nullptr)
+        { bfdebug_subbool(level, name, is_enabled(), msg); }
+    }
+
+    namespace apic_global_enable
+    {
+        constexpr const auto mask = 0x0000000000000800ULL;
+        constexpr const auto from = 11ULL;
+        constexpr const auto name = "apic_global_enable";
+
+        inline auto is_enabled()
+        { return is_bit_set(_read_msr(addr), from); }
+
+        inline auto is_enabled(value_type msr)
+        { return is_bit_set(msr, from); }
+
+        inline auto is_disabled()
+        { return is_bit_cleared(_read_msr(addr), from); }
+
+        inline auto is_disabled(value_type msr)
+        { return is_bit_cleared(msr, from); }
 
-        constexpr const auto disabled = 0x0ULL;
-        constexpr const auto invalid = 0x4ULL;
-        constexpr const auto xapic = 0x8ULL;
-        constexpr const auto x2apic = 0xCULL;
+        inline void enable()
+        { _write_msr(addr, set_bit(_read_msr(addr), from)); }
+
+        inline auto enable(value_type msr)
+        { return set_bit(msr, from); }
+
+        inline void disable()
+        { _write_msr(addr, clear_bit(_read_msr(addr), from)); }
+
+        inline auto disable(value_type msr)
+        { return clear_bit(msr, from); }
+
+        inline void dump(int level, std::string *msg = nullptr)
+        { bfdebug_subbool(level, name, is_enabled(), msg); }
+    }
+
+    namespace apic_base
+    {
+        constexpr const auto mask = 0xFFFFFFFFFFFFF000ULL;
```

Comment:
the mask should be from bit 12 to bit 35. Also lets go ahead and not shift the get and the set. That way the value returned is the proper address and the user doesn't have to remember to shift it back